deeppavlov/models/elmo/elmo.py

```diff
@@ -0,0 +1,456 @@
+            path.resolve()
+            log.info(f'[saving {epoch} epoch]')
+
+        path.parent.mkdir(parents=True, exist_ok=True)
+        path = str(path)
+
+        log.info(f'[saving model to {path}]')
+        with self.graph.as_default():
+            saver = tf.train.Saver()
+            saver.save(self.sess, path)
+
+    def train_on_batch(self,
+                       x_char_ids: list,
+                       y_token_ids: list) -> List[float]:
+        """
+        This method is called by trainer to make one training step on one batch.
+
+        Args:
+            x_char_ids:  a batch of char_ids
+            y_token_ids: a batch of token_ids
+
+        Returns:
+            value of loss function on batch
+        """
+
+        char_ids_batches, reversed_char_ids_batches = x_char_ids
+        token_ids_batches, reversed_token_ids_batches = y_token_ids
+
+        feed_dict = self._fill_feed_dict(char_ids_batches, reversed_char_ids_batches,
+                                         token_ids_batches, reversed_token_ids_batches)
+
+        with self.graph.as_default():
+            ret = self.sess.run([self.train_loss, self.train_op] + self.final_state_tensors, feed_dict)
+
+        self.init_state_values = ret[2:]
+        train_loss = ret[0]
+
+        return train_loss
+
+    def _build_model(self, train: bool, epoch: Optional[int] = None, **kwargs):
+
+        if hasattr(self, 'sess'):
+            self.sess.close()
+
+        self.options = copy.deepcopy(self.permanent_options)
+
+        if train:
+            self.options.update(self.train_options)
+            self.options.update(kwargs)
+
+            self.models, self.train_op, self.train_loss, _, self.graph = self._build_graph(tf.Graph())
+            self.loss = self.train_loss
+        else:
+            self.options.update(self.valid_options)
+            self.options.update(kwargs)
+
+            self.models, self.train_op, self.train_loss, self.loss, self.graph = self._build_graph(tf.Graph())
+
+        with self.graph.as_default():
+            self.init_state_values, self.init_state_tensors, self.final_state_tensors =\
+                self._init_session()
+        self.load(epoch)
+
+    def process_event(self, event_name, data):
+        if event_name == 'after_validation':
+            self._build_model(train = True)
+        elif event_name == 'after_epoch':
+            epoch = self.save_epoch_num + int(data['epochs_done'])
+            self.save(epoch)
+            self.save()
+            self.elmo_export(epoch)
+
+            self._build_model(train = False)
+
+    def elmo_export(self, epoch: Optional[int] = None) -> None:
+        """
+        Dump the trained weights from a model to a HDF5 file and export a TF-Hub module.
+        """
+        if hasattr(self, 'sess'):
+            self.sess.close()
+        path = self.save_path
+        if epoch:
+            from_path = path.parents[1] / self.epoch_save_path / str(epoch) / path.parts[-1]
```

Comment:
If my save_path is equal to `/home/user/elmo_model` then this code will save files somewhere in home directory.